CMakeLists.txt

```diff
@@ -322,6 +322,10 @@ endif()
 
 set(MIR_TRACEPOINT_LIB_INSTALL_DIR ${CMAKE_INSTALL_LIBDIR}/mir/tools)
 
+include(JoinPaths)
+join_paths(PKGCONFIG_LIBDIR "\${prefix}" "${CMAKE_INSTALL_LIBDIR}")
```

Comment:
Hm.

This doesn't actually do the right thing, does it? The path in pkg-config is expected to be relative to `prefix`, but if `CMAKE_INSTALL_LIBDIR` is set to an absolute path then `PKGCONFIG_LIBDIR` will *also* be set to an absolute path.

Reply:
It doesn't do "the right thing" in that it will substitute the placeholder with an absolute path. I'm not sure what the right thing *would* be though.

The situation in Nixpkgs (my downstream) is that package outputs live under separate prefixes, and headers may be installed to a separate `dev` output under a completely different prefix. So the passed-in paths might be:

- `CMAKE_INSTALL_PREFIX=/nix/store/<hash-A>-mir-<version>/`
- `CMAKE_INSTALL_INCLUDEDIR=/nix/store/<hash-B>-mir-<version>-dev/include`

Such a case will put both absolute paths as `prefix` and `includedir` into the pkg-config file. It could be possible to build a relative path from the `PREFIX` to the `INCLUDEDIR`, but this would look smth like

```pc
prefix=/nix/store/<hash-A>-mir-<version>/
includedir=${prefix}/../<hash-B>-mir-<version>-dev/include
```

which wouldn't make replacing the prefix any easier.

Reply:
Hmm, well `prefix` is inherently a public piece of information - exposed in the pkg-config file so the user may refer to & override it. But pretending that it holds any value when the user opted out of using the `prefix` by using separate absolute paths seems weird to me. If it mattered to me at configure time, then I would've used a relative path.

Maybe it could be turned relative if it can be determined that it lies within the `prefix`? CMake's package config generation helper [configure_package_config_file](https://cmake.org/cmake/help/v3.15/module/CMakePackageConfigHelpers.html#command:configure_package_config_file) employs such a tactic:

> The variables `<var1>` to `<varN>` given as `PATH_VARS` are the variables which contain install destinations. […] They are calculated by `configure_package_config_file` so that they are always relative to the installed location of the package. This works both for relative and also for absolute locations. For absolute locations it works only if the absolute location is a subdirectory of `INSTALL_PREFIX`.

Would you like me to make a custom function and copy this behaviour? This seems like an alright compromise between forcing a `prefix` where it doesn't belong, and taking all absolute path values as they are.

Reply:
I've gone ahead and added a `mir_make_pkgconfig_variable` function for this top-level use-case, let me know if this is any better. It takes the variable that should be set for the pkg-config substitution and the user-/default-supplied `CMAKE_INSTALL_<dir>` and will set the variable to
- `${prefix}/foo/bar` if the requested path lies within `CMAKE_INSTALL_PREFIX` so `prefix` can still be substituted, or
- `/absolute/foo/bar` if the path lies outside the `prefix` and would thus make dealing with `prefix` more cumbersome than just overriding the other variables.